homeassistant/components/calendar/trigger.py

```diff
@@ -41,34 +42,135 @@
 # mypy: disallow-any-generics
 
 
+@dataclass
+class QueuedCalendarEvent:
+    """An event that is queued to be fired in the future."""
+
+    trigger_time: datetime.datetime
+    event: CalendarEvent
+
+
+@dataclass
+class Timespan:
+    """A time range part of start/end dates, used for considering active events."""
+
+    start: datetime.datetime
+    """The start datetime of the interval."""
+
+    end: datetime.datetime
+    """The end datetime (exclusive) of the interval."""
+
+    def with_offset(self, offset: datetime.timedelta) -> Timespan:
+        """Return a new interval shifted by the specified offset."""
+        return Timespan(self.start + offset, self.end + offset)
+
+    def contains(self, trigger: datetime.datetime) -> bool:
+        """Return true if the trigger time is within the time span."""
+        return self.start <= trigger < self.end
+
+    def next_upcoming(
+        self, now: datetime.datetime, interval: datetime.timedelta
+    ) -> Timespan:
+        """Return a subsequent time span following the current time span.
+
+        This effectively gives us a cursor like interface for advancing through
+        time using the interval as a hint. The returned span may have a
+        different interval than the one specified.  For example, time span may
+        be longer during a daylight saving time transition, or may extend due to
+        drift if the current interval is old.  The returned time span is
+        adjacent and non-overlapping.
+        """
+        return Timespan(self.end, max(self.end, now) + interval)
+
+    def __str__(self) -> str:
+        """Return a compact string representation."""
+        return f"[{self.start}, {self.end})"
```

Comment:
Why is the opening bracket a square bracket and the closing bracket a round bracket?